### PR TITLE
Support "limited" enrollment access (field-level authorization)

### PR DESCRIPTION
**Pull request by gigxz**
## Description

Support "limited" enrollment access. This type of access lets you see the existence of the enrollment plus some limited information about it: Entry date, exit date, project/org name, project type, enrollment period, relationship to hoh, move-in date, and last bed night date.

I implemented this using field level authorization on the GQL Enrollment type. The majority of fields require the `can_view_enrollment_details` permission, but some fields can resolve with `can_view_limited_enrollment_details`.

Here is an example of the UI. This is an example client dashboard where the user only has permission to view enrollment details at Chestnut Grove organization, but is permitted to see limited enrollment details for enrollments at other projects.

https://www.pivotaltracker.com/n/projects/2591838/stories/186304198

Frontend PR: https://github.com/greenriver/hmis-frontend/pull/542

Screenshot 2023-10-30 at 5 19 08 PM

## Type of change
- [x] New feature (adds functionality)

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [ ] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb [hunk 1]
```diff
@@ -36,124 +42,134 @@ def self.configuration
     end
 
     description 'HUD Enrollment'
+
+    # LIMITED ACCESS FIELDS. These fields are visible with `can_view_limited_enrollment_details` permission (or `can_view_enrollment_details` + `can_view_project` perms).
```
Comment (gigxz) [inline drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb]: The `Enrollment` GraphQL type is now basically "split" - the majority of the fields require `can_view_enrollment_details` permissions, but a few fields don't.

The downside with this approach is that it is difficult to maintain. If someone adds a new field as a `field` or without specifying the required permission, it would be inadvertently exposed to users who should only have Limited access. (They wouldn't see the data in the application, but they could modify a query to fetch additional data).

I think another approach would be to resolve a union type - `LimitedEnrollment` and `Enrollment`. I looked into this a little bit but it seemed somewhat difficult to do with our pagination class. Happy to look into that further, or other ideas/patterns, if folks have any ideas.

File: drivers/hmis/app/graphql/types/base_field.rb [hunk 2]
```diff
@@ -19,6 +22,20 @@ def initialize(*args, default_value: nil, **kwargs, &block)
...
     end
 
+    # Field-level authorization
+    # https://graphql-ruby.org/authorization/authorization.html#field-authorization
+    def authorized?(object, args, ctx)
+      # if `permission:` was given, then require the current user to have the specified permissions on the object
+      base_authorized = super(object, args, ctx)
+      if @permissions.any?
+        base_authorized && @permissions.all? do |perm|
+          current_permission_for_context?(ctx, permission: perm, entity: object)
```
Comment (gigxz) [inline drivers/hmis/app/graphql/types/base_field.rb:32]: This uses the data loader to check permissions.

If the user lacks permission for a requested field, the field will be returned as null. (Its possible to change that behavior to throw an error, i'm on the fence)

File: drivers/hmis/app/graphql/types/hmis_schema/client.rb [hunk 3]
```diff
@@ -151,7 +151,13 @@ def external_ids
...
 
     def enrollments(**args)
-      resolve_enrollments(**args)
+      # If current user has "detailed" access to any enrollment for this client, then we also resolve
+      # "limited access" enrollments (if permitted). The purpose is to show additional enrollment history
+      # for "my" clients, but not for other clients in the system that I can see.
+      # This would need to change if we wanted to support the ability to see limited enrollment details for all clients.
+      has_some_detailed_access = current_permission?(permission: :can_view_enrollment_details, entity: object)
+      scope = object.enrollments.viewable_by(current_user, include_limited_access_enrollments: has_some_detailed_access)
+      resolve_enrollments(scope, **args, dangerous_skip_permission_check: true)
```
Comment (gigxz) [inline drivers/hmis/app/graphql/types/hmis_schema/client.rb:160]: this produces a lot of queries but I'm not sure how to reduce it. we need to find out if there are any detailed enrollments before we decide to include the limited enrollments

File: drivers/hmis/spec/requests/hmis/client_enrollments_performance_spec.rb [hunk 4]
```diff
@@ -105,9 +92,10 @@
 
     it 'minimizes n+1 queries' do
       expect do
-        _, result = post_graphql(**variables) { query }
+        response, result = post_graphql(**variables) { query }
+        expect(response.status).to eq(200), result.inspect
         expect(result.dig('data', 'client', 'enrollments', 'nodes').size).to eq(enrollments.size)
-      end.to make_database_queries(count: 10..30)
+      end.to make_database_queries(count: 10..35)
```
Comment (gigxz) [inline drivers/hmis/spec/requests/hmis/client_enrollments_performance_spec.rb:98]: This increased due to https://github.com/greenriver/hmis-warehouse/pull/3682/files#r1378003488 , I think. Would definitely like to fidn a way to bring it down

File: drivers/hmis/app/graphql/types/hmis_schema/client.rb (hunk 3 above)
Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/client.rb:160]: If the permissions can be checked on pre-loaded objects, you could possibly load all the enrollments for a given client (assuming that's the context) and preload the projects or whatever else you need to determine permission.  That might allow this to be only a handful of queries, but I agree, seems like this is moving toward a fairly heavy DB use, though maybe it's ok as long as we don't ever have hundreds of enrollments per client....

File: drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb (hunk 1 above)
Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb]: What about adding a new super method to use instead of `field` have `field` default to the more safe path, but to expose something to the limited users you'd have to call `limited_field` or something.  That way you have to be more intentional when exposing  data.

Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb]: Looks like you took the opposite approach and `detail_field` is the less safe path?  I'm a bit confused on what is being limited in this case, maybe a different name like "summary" or "public" would be easier to reason about?

File: drivers/hmis/app/models/hmis/role.rb [hunk 5]
```diff
@@ -236,6 +236,14 @@ def self.permissions_with_descriptions
...
         ],
       },
+      can_view_limited_enrollment_details: {
+        description: 'Grants access to view limited information about an enrollment, including: entry date, exit date, project name, project type, move-in date, and last bed night date.',
+        administrative: false,
+        access: [:viewable],
+        categories: [
+          'Enrollments',
+        ],
+      },
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/role.rb:246]: I'm starting to feel like this is what we intended the "warehouse" endpoint to handle.  Maybe time to revisit that.

File: drivers/hmis/app/graphql/types/hmis_schema/client.rb (hunk 3 above)
Comment (gigxz) [reply drivers/hmis/app/graphql/types/hmis_schema/client.rb:160]: Oh this only gets called once per client, not once per enrollment. It's "a lot of queries" (like maybe 10) just because our permission checks perform several queries, and we're doing 2 permission checks. This is the reason I had to bump the expected number of db queries in the test.

File: drivers/hmis/app/models/hmis/role.rb (hunk 5 above)
Comment (gigxz) [reply drivers/hmis/app/models/hmis/role.rb:246]: Is it? I always thought the 'warehouse endpoint' was meant to expose minimal data from _other_ data sources. But yeah, we should talk about plans for that going forward.

File: drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb (hunk 1 above)
Comment (gigxz) [reply drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb]: Great suggestions, thanks. I _should_ be able to override `field` just for this class. So we can have `field`(requiring enrollment details perm) and `summary_field`(?) for the others. I agree "Limited"/"Detail" are unclear here. For the permission I was mimicking the language from the warehouse (full client access vs limited client access).

Comment (gigxz) [reply drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb]: Ok, updated with https://github.com/greenriver/hmis-warehouse/pull/3682/commits/bfa0210d1b48581854bb1c080b3bd24eae54bd97. Much cleaner I think.

File: drivers/hmis/app/models/hmis/role.rb (hunk 5 above)
Comment (eanders) [reply drivers/hmis/app/models/hmis/role.rb:246]: As discussed offline, this is probably fine as is, happy to revisit the warehouse endpoint at a future date, especially since this still all falls within the HMIS data source.

**Review by eanders — APPROVED** [review]
Looks good to me!

File: drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb [hunk 6]
```diff
@@ -184,6 +214,21 @@ def project
       end
     end
 
+    # Needed because limited access viewers cannot resolve the project
+    def project_name
```
Comment (ttoomey) [inline drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb:218]: Interesting. Maybe there should be similar permissions to for orgs and projects to allow users to see just the summary fields; that way we can avoid spreading the fields onto enrollment. But I that seems like scope creep, this should be fine.

File: drivers/hmis/app/graphql/types/base_field.rb [hunk 7]
```diff
@@ -6,9 +6,12 @@
 
 module Types
   class BaseField < GraphQL::Schema::Field
+    include GraphqlPermissionChecker
     argument_class Types::BaseArgument
 
-    def initialize(*args, default_value: nil, **kwargs, &block)
+    def initialize(*args, default_value: nil, permission: nil, **kwargs, &block)
```
Comment (ttoomey) [inline drivers/hmis/app/graphql/types/base_field.rb]: suggestion: keyword could be `permissions: nil` (plural), that seems more accurate

File: drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb [hunk 8]
```diff
@@ -27,6 +27,26 @@ def self.configuration
...
+    end
+
+    # Add a separate "field" function to use for summary fields which are visible
+    # to users with "can_view_limited_enrollment_details".
+    #
+    # No field-level authorization is needed here, because the enrollment necessarily
+    # has some level of visibility if it's being resolved.
+    def self.summary_field(name, type = nil, **kwargs)
+      field(name, type, **kwargs, permission: nil)
+    end
```
Comment (ttoomey) [inline drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb]: This looks fine in terms of functionality. But I did want to float an idea for another approach here, particularly since this pattern might spread out into other resolvers. Feel free to ignore it if it seems like too much of a jump.

I know we have a pattern of creating `x_field` methods in our resolvers (such as `enrollments_field`, etc). There may be an opportunity to try a different approach here, particularly since this is a different dimension (authorization rather than data loading). 

The primary concern I have with the `x_field` pattern is that it creates a level of indirection that may be hard to understand. For example it may not be obvious that `assessments_field` will call `field` which we are overriding in this type, as opposed to passing it in explicitly with `assessments_field(permissions: xyz)`

I understand the current pattern addresses some important needs: preventing incorrect or missing permissions and also keeping our resolvers DRY. But maybe we could solve that in another way. One approach would be to change `field()` so that permissions are a required arg. Then we could use something like `with_options` to dry up the permissions declaration. Something like this:

```ruby
    # override field so permissions is required kwarg
    def self.field(name, type = nil, permissions:, **kwargs, &block)
      # See Types::BaseField `authorized?` function.
      super(name, type, permissions: permissions, **kwargs, &block)
    end

    # summary fields
    with_options permissions: [:can_view_enrollment_summary] do |type|
      type.field :household_id, ID, null: false
      type.assessments_field
      # etc
    end

    # detail fields
    with_options permissions: [:can_view_enrollment_details] do |type|
      type.field :household_id, ID, null: false
      # etc
    end

    # runtime error
    #  field :project, Types::HmisSchema::Project, null: false
    # this is okay
    #  field :project, Types::HmisSchema::Project, null: false, permissions: [:can_view_enrollment_details]


```

File: drivers/hmis/spec/requests/hmis/client_enrollments_performance_spec.rb (hunk 4 above)
Comment (ttoomey) [reply drivers/hmis/spec/requests/hmis/client_enrollments_performance_spec.rb:98]: It's just a few extra queries per client right? That doesn't seem so bad considering the granular permissions we are introducing.

**Review by ttoomey — APPROVED** [review]
Look great to me! Thank you for the clean up work in the specs!! 🙌 

I had one idea for a different approach for field level authorization but it shouldn't block this; the current implementation seems fine as is.

File: drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb (hunk 8 above)
Comment (gigxz) [reply drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb]: Nice, I like that approach. I agree that the indirection with overriding `field` is confusing, it's not particularly a pattern I'd like to spread to other resolvers.

I'm not _quite_ following how `type.assessments_field` would work in your example, but I like the direction. I'm not a huge fan of the current `HasAssessments` / `assessments_field` pattern.

I think for the sake of getting this out by end of week, I need to merge this and get it up for QA. But I think your suggestion would be great and I captured it here https://www.pivotaltracker.com/story/show/186384085! I don't have a strong sense of whether this type of field authorization will be needed in other places. There are 2 other instances I have in mind: (1) redacting SSN/DOB on the Client if user lacks access, and (2) redacting fields on File if its confidential and user lacks access. Those are both done in their own way, not using the ruby-graphql field authorization mechanism. They are much more contained than this Enrollment situation though.
